### PR TITLE
Include utility to migrate a CRD's storage version

### PR DESCRIPTION
**Pull request by dprotaso**
See https://github.com/knative/serving/pull/7499 for context

Essentially, the tool is generic enough it's possible to use on any CRD

### CODE REVIEW

File: apiextensions/storageversion/cmd/migrate/config.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+	cfg, err := getConfig(*masterURL, *kubeconfig)
+	if err != nil {
+		panic(fmt.Sprintf("Error building kubeconfig: %v", err))
+	}
+
+	return cfg
+}
+
+// GetConfig returns a rest.Config to be used for kubernetes client creation.
```
Comment (mattmoor) [inline apiextensions/storageversion/cmd/migrate/config.go:50]: nit: wrong name

File: apiextensions/storageversion/cmd/migrate/main.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		apixclient.NewForConfigOrDie(config),
+	)
+
+	ctx := signals.NewContext()
+
+	logger.Infof("Migrating %d group resources", len(grs))
+
+	for _, gr := range grs {
+		logger.Infof("Migrating group resource %s", gr)
+		if err := migrator.Migrate(ctx, gr); err != nil {
```
Comment (mattmoor) [inline apiextensions/storageversion/cmd/migrate/main.go:54]: use an `errgroup`?

File: apiextensions/storageversion/cmd/migrate/config.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// TODO - make a package that's reusable here and by sharedmain
```
Comment (mattmoor) [inline apiextensions/storageversion/cmd/migrate/config.go:30]: ☝️ 👍

File: apiextensions/storageversion/cmd/migrate/main.go (hunk 2 above)
Comment (dprotaso) [reply apiextensions/storageversion/cmd/migrate/main.go:54]: `errgroup` adds parallelism and I didn't feel the need to optimize given the rate limiting within the clients. Secondly, the k8s migrator tool tries to reduce the load on the API server. 

I'd say let's revisit this if there's a need.

File: apiextensions/storageversion/migrator.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		return fmt.Errorf("unable to fetch crd %s - %w", gr, err)
+	}
+
+	version := storageVersion(crd)
+
+	if err := m.migrateResources(ctx, gr.WithVersion(version)); err != nil {
+		return err
+	}
+
+	patch := fmt.Sprintf(`{"status":{"storedVersions":["` + version + `"]}}`)
```
Comment (vagababov) [inline apiextensions/storageversion/migrator.go:59]: you're not actually springing anything :) So just
```suggestion
	patch :=`{"status":{"storedVersions":["` + version + `"]}}`
```

Comment (dprotaso) [reply apiextensions/storageversion/migrator.go:59]: lol

File: apiextensions/storageversion/cmd/migrate/main.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			logger.Fatalf("Failed to migrate: %s", err)
+		}
+	}
+
+	logger.Infof("Migration complete")
```
Comment (vagababov) [inline apiextensions/storageversion/cmd/migrate/main.go:59]:
```suggestion
	logger.Info("Migration complete")
```

File: apiextensions/storageversion/migrator_test.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+		{
+			name: "failed to fetch CRD",
+			crd: func(fake *k8stesting.Fake) {
+				fake.PrependReactor("get", "*",
+					func(k8stesting.Action) (bool, runtime.Object, error) {
+						return true, nil, fmt.Errorf("failed to get crd")
+					})
+			},
+		},
+		{
```
Comment (vagababov) [inline apiextensions/storageversion/migrator_test.go:115]:
```suggestion
	}{{
			name: "failed to fetch CRD",
			crd: func(fake *k8stesting.Fake) {
				fake.PrependReactor("get", "*",
					func(k8stesting.Action) (bool, runtime.Object, error) {
						return true, nil, fmt.Errorf("failed to get crd")
					})
			},
		}, {
```
Same with others.

File: apiextensions/storageversion/migrator_test.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+						return true, nil, fmt.Errorf("failed to patch definition")
+					})
+			},
+		},
+		// todo paging fails
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
```
Comment (vagababov) [inline apiextensions/storageversion/migrator_test.go:147]:
```suggestion
```

File: apiextensions/storageversion/migrator_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+				test.crd(&cclient.Fake)
+			}
+
+			if test.dyn != nil {
+				test.dyn(&dclient.Fake)
+			}
+
+			m := NewMigrator(dclient, cclient)
+			if err := m.Migrate(context.TODO(), fakeGR); err == nil {
+				t.Errorf("Migrate should have returned an error")
```
Comment (vagababov) [inline apiextensions/storageversion/migrator_test.go:162]:
```suggestion
				t.Error("Migrate should have returned an error")
```

**Pull request merged**